st2reactor/st2reactor/container/process_container.py

```
@@ -61,65 +76,109 @@ def __init__(self, sensors, poll_interval=5):
+                eventlet.sleep(self._poll_interval)
         except greenlet.GreenletExit:
             # This exception is thrown when sensor container manager
             # kills the thread which runs process container. Not sure
             # if this is the best thing to do.
-            self.stopped = True
+            self._stopped = True
             return SUCCESS_EXIT_CODE
         except:
             LOG.exception('Container failed to run sensors.')
-            self.stopped = True
+            self._stopped = True
             return FAILURE_EXIT_CODE
 
-        self.stopped = True
+        self._stopped = True
         LOG.error('Process container quit. It shouldn\'t.')
+        return SUCCESS_EXIT_CODE
 
     def _poll_sensors_for_results(self, sensor_ids):
+        """
+        Main loop which polls sensor for results and detects dead sensors.
+        """
         for sensor_id in sensor_ids:
+            now = int(time.time())
+
             process = self._processes[sensor_id]
             status = process.poll()
 
             if status is not None:
                 # Dead process detected
-                LOG.info('Process for sensor %s has exited with code %s',
-                         self._sensors[sensor_id]['ref'], status)
+                LOG.info('Process for sensor %s has exited with code %s', sensor_id, status)
+
                 sensor = self._sensors[sensor_id]
+                self._delete_sensor(sensor_id)
+
                 self._dispatch_trigger_for_sensor_exit(sensor=sensor,
                                                        exit_code=status)
-                self._delete_sensors(sensor_id)
+
+                # Try to respawn a dead process (maybe it was a simple failure which can be
+                # resolved with a restart)
+                self._respawn_sensor(sensor_id=sensor_id, sensor=sensor, exit_code=status)
+            else:
+                sensor_start_time = self._sensor_start_times[sensor_id]
```

It won't mask because you'd still log the sensor_id which doesn't exist and you can raise an exception :P. Anyways, you have my suggestion. I can always say "I told you" :P.